### PR TITLE
Shoot pod/service CIDR defaults per seed

### LINKED ISSUE #895: Default shoot networks should be configured on seed side
Currently, the default pod/service networks for shoots are hard-coded in the Garden API: https://github.com/gardener/gardener/blob/master/pkg/apis/core/v1alpha1/types_utils.go#L91-L96

This means that every shoot that does not specify a pod/service network will get these values. However, due to the VPN we install between seed and shoot the shoot networks must be disjunct from the pod/service network of the seed. That means we can't have these static defaults for all shoots. Instead, we should enable the possibility to maintain the defaults on the seed resource itself:

```yaml
apiVersion: garden.sapcloud.io/v1beta1
kind: Seed
...
spec:
  networks: # Seed and Shoot networks must be disjunct
    nodes: 10.240.0.0/16
    pods: 10.241.128.0/17
    services: 10.241.0.0/17
    shootDefaults: # or similar
      pods: 100.96.0.0/11
      services: 100.64.0.0/13
```

The static defaulting [here](https://github.com/gardener/gardener/blob/master/pkg/apis/garden/v1beta1/defaults.go#L40-L43) should be replaced by a dynamic defaulting via admission, preferably the `SeedManager` admission controller takes over.

Reported by @mandelsoft and @MartinWeindel.

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**:
Specify the shoot pod/service default CIDR per seeds.

**Which issue(s) this PR fixes**:
Fixes #895

**Special notes for your reviewer**:
Requirement for #1376

**Release note**:

```noteworthy operator
It is now possible to specify shoot default pod/service CIDRs **per seed**. Please configure `spec.networks.shootDefaults` on your `Seed` resources if you want to use this.
```

### CODE REVIEW

File: pkg/apis/garden/v1beta1/defaults.go [hunk 1]
```diff
@@ -241,6 +185,33 @@ func SetDefaults_Seed(obj *Seed) {
...
+		defaultServiceCIDR         = DefaultServiceNetworkCIDR
+		defaultPodCIDRAlicloud     = DefaultPodNetworkCIDRAlicloud
+		defaultServiceCIDRAlicloud = DefaultServiceNetworkCIDRAlicloud
+	)
+
+	if obj.Spec.Networks.ShootDefaults == nil {
+		obj.Spec.Networks.ShootDefaults = &ShootNetworks{}
+	}
+
+	if v, ok := obj.Annotations[garden.MigrationSeedProviderType]; ok && v == "alicloud" {
```
Comment (timuthy) [inline pkg/apis/garden/v1beta1/defaults.go:200]: Please check that you assign the `defaultPodCIDRAlicloud ` and `defaultServiceCIDRAlicloud` here.

File: pkg/apis/garden/v1beta1/defaults.go [hunk 2]
```diff
@@ -241,6 +185,33 @@ func SetDefaults_Seed(obj *Seed) {
...
+		if obj.Spec.Networks.ShootDefaults.Pods == nil {
+			obj.Spec.Networks.ShootDefaults.Pods = &defaultPodCIDR
+		}
+		if obj.Spec.Networks.ShootDefaults.Services == nil {
+			obj.Spec.Networks.ShootDefaults.Services = &defaultServiceCIDR
+		}
+	} else {
```
Comment (timuthy) [inline pkg/apis/garden/v1beta1/defaults.go:207]: Please check that you assign the `defaultPodCIDR ` and `defaultServiceCIDR` here.

**Review by timuthy — CHANGES_REQUESTED** [review]

**Comment by timuthy** [comment]
Can you also generate the example manifests and check them in?

File: pkg/apis/garden/v1beta1/defaults.go (hunk 1 above)
Comment (rfranzke) [reply pkg/apis/garden/v1beta1/defaults.go:200]: nice catch!

**Review by timuthy — DISMISSED** [review]
/lgtm

File: pkg/apis/core/v1alpha1/types_seed.go [hunk 3]
```diff
@@ -117,6 +117,19 @@ type SeedNetworks struct {
 	Pods CIDR `json:"pods"`
 	// Services is the CIDR of the service network.
 	Services CIDR `json:"services"`
+	// ShootDefaults contains the default networks CIDRs for shoots.
+	// +optional
+	ShootDefaults *ShootNetworks `json:"shootDefaults,omitempty"`
+}
+
+// ShootDefaults contains the default networks CIDRs for shoots.
```
Comment (ialidzhikov) [inline pkg/apis/core/v1alpha1/types_seed.go]:
```suggestion
// ShootNetworks contains the default networks CIDRs for shoots.
```

File: pkg/apis/garden/v1beta1/types.go [hunk 4]
```diff
@@ -1870,3 +1883,14 @@ type BackupInfrastructureStatus struct {
...
 }
+
+const (
+	// DefaultPodNetworkCIDR is a constant for the default pod network CIDR of a Shoot cluster.
+	DefaultPodNetworkCIDR = gardencorev1alpha1.CIDR("100.96.0.0/11")
+	// DefaultServiceNetworkCIDR is a constant for the default service network CIDR of a Shoot cluster.
+	DefaultServiceNetworkCIDR = gardencorev1alpha1.CIDR("100.64.0.0/13")
+	// DefaultPodNetworkCIDRAlicloud is a constant for the default pod network CIDR of a Alicloud Shoot cluster.
+	DefaultPodNetworkCIDRAlicloud = gardencorev1alpha1.CIDR("100.64.0.0/11")
+	// DefaultServiceNetworkCIDR is a constant for the default service network CIDR of a Alicloud Shoot cluster.
```
Comment (ialidzhikov) [inline pkg/apis/garden/v1beta1/types.go]:
```suggestion
      // DefaultServiceNetworkCIDRAlicloud is a constant for the default service network CIDR of a Alicloud Shoot cluster.
```

**Review by ialidzhikov — COMMENTED** [review]
/lgtm

**Pull request merged**